modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/GridNearTxFinishFuture.java

```diff
@@ -978,7 +982,27 @@ public GridDistributedTxMapping mapping() {
 
         /** {@inheritDoc} */
         @Override boolean onNodeLeft(UUID nodeId, boolean discoThread) {
+            if (tx.state() == COMMITTING || tx.state() == COMMITTED) {
+                Map<UUID, Collection<UUID>> txNodes = tx.transactionNodes();
+
+                Collection<UUID> backups = txNodes.get(nodeId);
+
+                boolean hasBackups = false;
+
+                if (backups != null)
+                    hasBackups = backups.stream().anyMatch(backupId -> cctx.discovery().node(backupId) != null);
+
+                if (cctx.discovery().node(m.primary().id()) == null && !hasBackups) {
+                    onDone(new TransactionHeuristicException("Primary node [nodeId=" + nodeId + ", consistentId=" +
```

Comment:
You should print informative partition loss message here, like here:
org.apache.ignite.internal.processors.cache.distributed.dht.GridDhtTopologyFutureAdapter#validateKey.

new CacheInvalidStateException("Failed to commit a transaction (all partition owners have left the grid, partition data has been lost) [cacheName=cache, partition=10, key=somekey]")

I think better to avoid heuristic exception and use CacheException instread.